frontend/src/pages/TermPlanner/ContextMenu/ContextMenu.jsx

```diff
@@ -41,21 +45,92 @@ const ContextMenu = ({ code, plannedFor }) => {
     dispatch(addTab(code));
   };
 
+  const [isEditMarkVisible, setIsEditMarkVisible] = useState(false);
+
+  const showEditMark = () => {
+    setIsEditMarkVisible(true);
+  };
+
+  const validLetterGrades = ["SY", "PS", "CR", "DN", "HD"];
+  const [markInputBuf, setMarkInputBuf] = useState(
+    useSelector((state) => state.planner.courses[code].mark),
+  );
+
+  const handleConfirmEditMark = (mark) => {
+    const attemptedMark = (` ${mark}`).slice(1).replace(" ", "");
+    if (
+      (attemptedMark.isNaN && !validLetterGrades.includes(attemptedMark))
+      || (parseFloat(attemptedMark) < 0 || parseFloat(attemptedMark) > 100)
+    ) {
+      return message.error("Could not update mark. Please enter a valid mark or letter grade");
+    }
+    dispatch(updateCourseMark({
+      code,
+      mark: attemptedMark,
+    }));
+    setIsEditMarkVisible(false);
+
+    validateTermPlanner(
+      dispatch,
+      { years, startYear, completedTerm },
+      { programCode, majors, minor },
+    );
+
+    return message.success("Mark Updated");
+  };
+
+  const [markUpdatedQueued, setMarkUpdateQueued] = useState(false);
+  useEffect(() => {
+    if (markUpdatedQueued) {
+      setMarkUpdateQueued(false);
+      handleConfirmEditMark(markInputBuf);
+    }
+  });
+
+  const queueEditMark = () => {
+    setMarkUpdateQueued(true);
+  };
+
+  const handleCancelEditMark = () => {
+    setIsEditMarkVisible(false);
+    setMarkUpdateQueued(false);
+  };
+
   return (
-    <Menu id={id} theme={theme.dark}>
-      {plannedFor && (
-        <Item onClick={handleUnschedule}>
-          <FaRegCalendarTimes className="context-menu-icon" /> Unschedule
+    <>
+      <Menu id={id} theme={theme.dark}>
+        {plannedFor && (
+          <Item onClick={handleUnschedule}>
+            <FaRegCalendarTimes className="context-menu-icon" /> Unschedule
+          </Item>
+        )}
+        <Item onClick={handleDelete}>
+          <DeleteFilled className="context-menu-icon" /> Delete from Planner
+        </Item>
+        <Item onClick={showEditMark}>
+          <FaEdit className="contextMenuIcon" /> Edit mark
+        </Item>
+        <Item onClick={handleInfo}>
+          <InfoCircleFilled className="context-menu-icon" />
+          View Info
         </Item>
-      )}
-      <Item onClick={handleDelete}>
-        <DeleteFilled className="context-menu-icon" /> Delete from Planner
-      </Item>
-      <Item onClick={handleInfo}>
-        <InfoCircleFilled className="context-menu-icon" />
-        View Info
-      </Item>
-    </Menu>
+      </Menu>
+      <Modal
+        title={`Edit Mark: ${code}`}
+        visible={isEditMarkVisible}
+        onOk={handleConfirmEditMark}
+        onCancel={handleCancelEditMark}
+        width="300px"
+      >
+        <EditMarks
+          courseCode={code}
+          inputBuffer={markInputBuf}
+          letterGrades={validLetterGrades}
+          setInputBuffer={setMarkInputBuf}
+          handleConfirm={queueEditMark}
```

Comment:
I'm a bit confused on why we are indirectly calling `handleConfirmEditMark` through `queueEditMark` mark? Imo adds a bit of complexity. Couldn't we just directly call `handleConfirmEditMark` instead?

Reply:
yup now this is more sane

Reply:
Wait I can't read

Reply:
> I'm a bit confused on why we are indirectly calling `handleConfirmEditMark` through `queueEditMark` mark? Imo adds a bit of complexity. Couldn't we just directly call `handleConfirmEditMark` instead?

If you take the following
```jsx
const [var, setVar] = useState();

return <Component var={var} setVar={setVar} />
```
If the child component wants to call `setVar` and then callback a function from the parent like `handleConfirmEditMark`, that function will actually still use the old value of `var` because it doesn't get updated till the next reload which means, we will dispatch an older version of the input.

This isn't really an issue when the user types in input, since the act of typing in input (which is what sets the buffer) and, confirming your input, are two separate actions. However, when it comes to `letterGrade`, we want to save the user a click by cutting out the need to press `confirm` after clicking on a grade.

However, if we simply call `handleConfirmEditMark` immediately, the `ContextMenu` will not have refreshed and we will dispatch the old value already in the input. This will make it so that u actually need to click twice on the `letterGrade` to successfully update the mark.

We can avoid this by having all `confirm` actions signal that the mark is ready to be updated. Then, as soon as `ContextMenu` refreshes due to this, the mark will be dispatched and the `EditMarks` menu will be closed on success.

@Peedee2002 This is the queue I was referring to on Sunday.
 @gtangelo

Reply:
Okay I now get what you're trying to do @imagine-hussain. You're approach is definitely valid! I guess what confused me was the fact that there was alot of one liner functions (which may or may not add a bit of bloat to the code) that can be just moved to as anonymous functions instead. Another thing that confused me was the variable name for `inputBuffer` (nothing wrong with the name itself, just never seen it named as a 'buffer' for an input value before lol - usually just seen it named more explicit i.e. in this case, somthing like `mark` or `markVal`.


> I think it would be better to encapsulate the logic of the edit marks modal within a component alongside EditMarks? Rather than making EditMarks a seperate component, instead, make the edit modal component as it's own component? This way we can encasulate all associated logic (i.e. states, handle functions, etc.) inside that component file.

I guess a more straightforward way to implement to `EditMarks` component would be something like this below (this is very rough code so it may need some improving if u want to use it lol)? This is also in relation to my comment made in that we should have the `EditMarksModal` component which encapsulates the whole editing marks logic.

```
const EditMarksModal = ({
  code, visible, setVisible,
}) => {
  const [mark, setMark] = useState(useSelector((state) => state.planner.courses[code].mark));
  const dispatch = useDispatch();
  const letterGrades = ["FL", "PS", "CR", "DN", "HD"];

  const handleConfirmEditMark = () => {
    if (
      (mark.isNaN && !letterGrades.includes(mark))
      || (parseFloat(mark) < 0 || parseFloat(mark) > 100)
    ) {
      return message.error("Could not update mark. Please enter a valid mark or letter grade");
    }
    dispatch(updateCourseMark({
      code,
      mark,
    }));
    setVisible(false);
    return message.success("Mark Updated");
  };

  const handleEnter = (e) => {
    if (e.key === "Enter") {
      handleConfirmEditMark();
    }
  };

  const handleConfirmLetterGradeMark = (grade) => {
    dispatch(updateCourseMark({
      code,
      mark: grade,
    }));
    setMark(grade);
    setVisible(false);
    return message.success("Mark Updated");
  };

  return (
    <Modal
      title={`Edit Mark: ${code}`}
      visible={visible}
      onOk={handleConfirmEditMark}
      onCancel={() => setVisible(false)}
      width="300px"
    >
      <div className="edit-mark">
        <div className="edit-mark-head" />
        <Input
          onChange={(e) => setMark(e.target.value)}
          onKeyDown={handleEnter}
          placeholder="Enter Mark"
          value={mark}
        />
        <div className="letter-grade-container">
          {
            letterGrades.map((letterGrade) => (
              <Button onClick={() => handleConfirmLetterGradeMark(letterGrade)} className="letter-grade-button">
                {letterGrade}
              </Button>
            ))
          }
        </div>
      </div>
    </Modal>
  );
};
```